packages/worker/src/api/controllers/global/users.ts

```diff
@@ -22,6 +22,65 @@ export const save = async (ctx: any) => {
   }
 }
 
+export const bulkSave = async (ctx: any) => {
+  let { users: newUsersRequested, groups } = ctx.request.body
+  let groupsToSave: any[] = []
+  const newUsers: any[] = []
+  const db = tenancy.getGlobalDB()
+  const currentUserEmails =
+    (await users.allUsers())?.map((x: any) => x.email) || []
+  for (const newUser of newUsersRequested) {
+    if (
+      newUsers.find((x: any) => x.email === newUser.email) ||
+      currentUserEmails.includes(newUser.email)
+    )
+      continue
+    newUser.userGroups = groups
+    newUsers.push(newUser)
+  }
+
+  if (groups.length) {
+    groups.forEach(async (groupId: string) => {
+      let oldGroup = await db.get(groupId)
+      groupsToSave.push(oldGroup)
+    })
+  }
+
+  try {
+    let response: any[] = []
+    for (const user of newUsers) {
+      response.push(
+        await users.save(user, {
```

Comment:
Ideally, this bulk save functionality could be built with the `bulkDocs` API - https://pouchdb.com/guides/bulk-operations.html, if you are updating say 100 users here this is very inefficient as lots of individual gets and writes, vs extracting everything you need upfront, building all the docs and then inserting them all in a single fast operation.

Just thinking if someone comes along and tries to upload their entire companies worth of users, they'll run into problems/very slow response times, it might feel broken.

A method to achieve this could be breaking out the synchronous component of the `users.save` core functionality, calling it to gather up all the docs you need to write, getting them into an array and then calling the `bulkDocs` functionality after this for loop.

Reply:
We need to be very careful around this WRT events, in the current `bulkCreate` impl it will bypass user save events as we effectively use the save as a validator / object generator instead and defer the save. The events are used for
- analytics 
- licensing

It'll also skip user uniqueness checks. I think we should remove this flag as it's totally unsafe and interferes with application logic in a few ways. 

There's an option for this going forward

**Asynchronous saves**
- Return the bulk api call immediately
- Process saves as a background job, one by one
- Notify the user when the job is done

This relies on what we have been calling just 'background jobs' for our application, which still needs spec'd out and implemented properly. The only other potential option would be something like

**Synchronous saves with back populated events**
- Refactor the `save` logic to have a shared `newUser` function
- Use this in both regular `save` calls as well as a `saveBulk` implementation
- Retroactively make sure that we handle events after the db save has been done in the bulk impl

However it still have a background job element, just not something that needs to be communicated to the user.
I don't think it's feasible though, especially in cloud as we need to perform these uniqueness checks:
- check budibase users inside the tenant
- check budibase users in other tenants
- check root account users in account portal

Skipping this even temporarily could be very bad. 

I think until we have background jobs, handling user saves will need to be a synchronous operation. In future we can detect if size is greater than X and create the user in batches in the background.

Reply:
I pushed some changes just as you commented Rory, is this closer to what you meant in regards to the synchronous saves?

Reply:
The approach you've put in peter is pretty much bang on the second option peter 👍 

Yeah you have called out the main issue here michael, the account check is the inefficient part here. Going forward as a part of the active user pricing there is also now a user representation in the account portal dynamo table. Whenever we create a user there will be a minimal representation there, and when we delete a user we will be setting a `deletedAt` attribute. These will be run as a part of the user created and user deleted events so they will be covered in this implementation, but the advantage gained here in terms of fast response time with a bulk write is now gone as we will be using a network call to account portal even if we are skipping the root account exists checks and looping db inserts. 

Traditionally a microservice architecture would use a message queue for this, where we could do the creates very quickly up front and dump all the users on a queue for dependant processing in account portal. This is more or less the same approach used for what we will call background jobs. Still there's no getting away from the uniqueness check up front here. We could in theory remove the account uniqueness check, as the budibase user would also be created in the db and should meet the points you said about checking up front michael, however there are still some issues:

- Self hosted account email address could be used for a cloud internal user (breaks login)
- Unverified cloud account user added to different tenant as internal user. This isn't something we've tested but will for sure mess with tenancy when the user verifies. Would be a common scenario when a few new people are trying to get off the ground with budibase and are creating accounts, trying to add each other to their accounts etc. 

We're in an unfortunate situation around multiple user stores. If we had a single source of truth for all users which would be a more common scenario we wouldn't have problems like this